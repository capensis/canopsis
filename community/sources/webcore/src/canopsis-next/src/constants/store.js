export const ENTITIES_TYPES = {
  alarm: 'alarm',
  alarmDetails: 'alarmDetails',
  alarmDetailsSteps: 'alarmDetailsSteps',
  entity: 'entity',
  service: 'entityservices',
  weatherService: 'weatherService',
  pbehavior: 'pbehavior',
  event: 'event',
  userPreference: 'userPreference',
  group: 'group',
  view: 'view',
  viewTab: 'viewTab',
  widget: 'widget',
  widgetTemplate: 'widgetTemplate',
  stat: 'stat',
  user: 'user',
  role: 'role',
  eventFilter: 'eventFilter',
  metaAlarmRule: 'metaAlarmRule',
  snmpRule: 'snmpRule',
  dynamicInfo: 'dynamicInfo',
  broadcastMessage: 'broadcastMessage',
  playlist: 'playlist',
  pbehaviorExceptions: 'pbehaviorExceptions',
  pbehaviorTypes: 'pbehaviorTypes',
  pbehaviorReasons: 'pbehaviorReasons',
  remediationInstruction: 'remediationInstruction',
  remediationJob: 'remediationJob',
  remediationConfiguration: 'remediationConfiguration',
  remediationInstructionExecution: 'remediationInstructionExecution',
  remediationInstructionStats: 'remediationInstructionStats',
  scenario: 'scenario',
  entityCategory: 'entityCategory',
  testSuite: 'testSuite',
  testSuiteHistory: 'testSuiteHistory',
  stateSetting: 'stateSetting',
  idleRules: 'idleRules',
  flappingRules: 'flappingRules',
  resolveRules: 'resolveRules',
  filter: 'filter',
  ratingSettings: 'ratingSettings',
  pattern: 'pattern',
  map: 'map',
  alarmTag: 'alarmTag',
  shareToken: 'shareToken',
  declareTicketRule: 'declareTicketRule',
};
