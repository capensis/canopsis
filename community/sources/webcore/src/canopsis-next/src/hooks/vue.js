import { getCurrentInstance } from 'vue';

/**
 * Hook for get current component instance
 *
 * @return {Vue}
 */
export const useComponentInstance = () => {
  const instance = getCurrentInstance();

  return instance?.proxy || instance;
};

/**
 * Hook for get current component options
 *
 * @return {Object}
 */
export const useComponentOptions = () => {
  const instance = useComponentInstance();

  return instance.$options;
};

/**
 * Hook for get current component model
 *
 * @return {Object}
 */
export const useComponentModel = () => {
  const { model } = useComponentOptions();

  return {
    event: model?.event ?? 'input',
    prop: model?.event ?? 'value',
  };
};
