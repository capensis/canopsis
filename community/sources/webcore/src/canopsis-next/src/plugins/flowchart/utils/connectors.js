import { range } from 'lodash';

import uid from '@/helpers/uid';

/**
 * @typedef {Object} ConnectorPoint
 * @property {number} x
 * @property {number} y
 * @property {string} _id
 */

/**
 * Generate connector
 *
 * @param {number} x
 * @param {number} y
 * @returns {ConnectorPoint}
 */
export const generateConnectorPoint = (x, y) => ({
  x,
  y,
  _id: uid(),
});

/**
 * Generate connector points by rectangle
 *
 * @param {Rect} rect
 * @param {number} count
 * @returns {ConnectorPoint[]}
 */
export const generateConnectorsByRect = (rect, count) => {
  const { width, height, rx, ry, x, y } = rect;
  const sideConnectorsCount = Math.round(count / 4);
  const sideSpacesCount = sideConnectorsCount + 1;
  const startX = x + rx;
  const startY = x + ry;

  const widthConnectorOffset = (width - rx * 2) / (sideSpacesCount);
  const heightConnectorOffset = (height - ry * 2) / (sideSpacesCount);

  const sideConnectors = range(1, sideConnectorsCount + 1);

  const topConnectors = sideConnectors.map(index => generateConnectorPoint(
    startX + widthConnectorOffset * index,
    y,
  ));

  const rightConnectors = sideConnectors.map(index => generateConnectorPoint(
    x + width,
    startY + heightConnectorOffset * index,
  ));

  const bottomConnectors = sideConnectors.map(index => generateConnectorPoint(
    startX + widthConnectorOffset * index,
    y + height,
  ));

  const leftConnectors = sideConnectors.map(index => generateConnectorPoint(
    x,
    startY + heightConnectorOffset * index,
  ));

  const connectors = [
    ...topConnectors,
    ...rightConnectors,
    ...bottomConnectors,
    ...leftConnectors,
  ];

  if (!rx && !ry) {
    /* Angles connectors */
    connectors.push(
      generateConnectorPoint(x, y),
      generateConnectorPoint(x + width, y),
      generateConnectorPoint(x + width, y + height),
      generateConnectorPoint(x, y + height),
    );
  }

  return connectors;
};
