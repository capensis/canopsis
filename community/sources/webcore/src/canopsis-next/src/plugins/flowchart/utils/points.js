/**
 * Calculate points with ghosts points
 *
 * @param {Array} points
 * @returns {Array}
 */
export const getPointsWithGhosts = points => points.reduce((acc, point, index) => {
  const nextIndex = index + 1;
  const nextPoint = points[nextIndex];

  acc.push({ ...point, index });

  if (nextPoint) {
    acc.push({
      x: (point.x + nextPoint.x) / 2,
      y: (point.y + nextPoint.y) / 2,
      ghost: true,
      index: nextIndex,
    });
  }

  return acc;
}, []);
