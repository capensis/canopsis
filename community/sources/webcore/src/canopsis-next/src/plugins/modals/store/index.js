import Vue from 'vue';

import { VUETIFY_ANIMATION_DELAY } from '@/config';

import uid from '@/helpers/uid';

export const types = {
  SHOW: 'SHOW',
  HIDE: 'HIDE',
  HIDE_COMPLETED: 'HIDE_COMPLETED',
  MINIMIZE: 'MINIMIZE',
  MAXIMIZE: 'MAXIMIZE',
};

export default {
  namespaced: true,
  state: {
    allIds: [],
    byId: {},
  },
  getters: {
    modals: state => state.allIds.map(id => state.byId[id]),
    hasMaximizedModal: state => state.allIds.some(id => !state.byId[id].minimized),
  },
  mutations: {
    [types.SHOW](state, {
      id,
      name,
      config = {},
      dialogProps = {},
    }) {
      Vue.set(state.byId, id, {
        id,
        name,
        config,
        dialogProps,
        hidden: false,
        minimized: false,
      });

      state.allIds.push(id);
    },
    [types.HIDE](state, { id }) {
      Vue.set(state.byId[id], 'hidden', true);
    },
    [types.HIDE_COMPLETED](state, { id }) {
      state.allIds = state.allIds.filter(value => value !== id);

      Vue.delete(state.byId, id);
    },
    [types.MINIMIZE](state, { id }) {
      Vue.set(state.byId[id], 'minimized', true);
    },
    [types.MAXIMIZE](state, { id }) {
      Vue.set(state.byId[id], 'minimized', false);
    },
  },
  actions: {
    /**
     * Show modal window by name with unique id
     *
     * @param {Function} commit
     * @param {Object} state
     * @param {string} name
     * @param {Object} [config = {}]
     * @param {Object} [dialogProps = {}]
     * @param {string} [id = uid()]
     */
    show({ commit, state }, {
      name,
      config = {},
      dialogProps = {},
      id = uid('modal'),
    } = {}) {
      if (state.byId[id]) {
        return commit(types.MAXIMIZE, { id });
      }

      return commit(types.SHOW, {
        id,
        name,
        config,
        dialogProps,
      });
    },

    /**
     * Hide modal by id
     *
     * @param {Function} commit
     * @param {Object} state
     * @param {string} [id]
     */
    hide({ commit, state }, { id } = {}) {
      if (!id) {
        throw new Error('Missed required parameter');
      }

      if (!state.byId[id]) {
        return;
      }

      commit(types.HIDE, { id });

      /**
       * This function added for vuetify animation waiting
       */
      setTimeout(() => {
        if (state.byId[id]) {
          commit(types.HIDE_COMPLETED, { id });
        }
      }, VUETIFY_ANIMATION_DELAY);
    },
    /**
     * Minimize modal by id
     *
     * @param {Function} commit
     * @param {string} [id]
     */
    minimize({ commit }, { id } = {}) {
      if (!id) {
        throw new Error('Missed required parameter');
      }

      commit(types.MINIMIZE, { id });
    },

    /**
     * Maximize modal by id
     *
     * @param {Function} commit
     * @param {string} [id]
     */
    maximize({ commit }, { id } = {}) {
      if (!id) {
        throw new Error('Missed required parameter');
      }

      commit(types.MAXIMIZE, { id });
    },
  },
};
