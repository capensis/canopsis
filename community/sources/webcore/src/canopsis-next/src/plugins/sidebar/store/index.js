import { VUETIFY_ANIMATION_DELAY } from '@/config';

export const types = {
  SHOW: 'SHOW',
  HIDE: 'HIDE',
  HIDE_COMPLETED: 'HIDE_COMPLETED',
};

export default {
  namespaced: true,
  state: {
    name: null,
    config: {},
    hidden: true,
  },
  getters: {
    sidebar: state => state,
  },
  mutations: {
    [types.SHOW](state, { name, config = {} }) {
      state.name = name;
      state.config = config;
      state.hidden = false;
    },
    [types.HIDE](state) {
      state.hidden = true;
    },
    [types.HIDE_COMPLETED](state) {
      state.name = null;
      state.config = {};
    },
  },
  actions: {
    show({ commit }, { name, config = {} } = {}) {
      commit(types.SHOW, { name, config });
    },

    hide({ commit, state }) {
      commit(types.HIDE);

      /**
       * This function added for vuetify animation waiting
       */
      setTimeout(() => {
        if (state.hidden) {
          commit(types.HIDE_COMPLETED);
        }
      }, VUETIFY_ANIMATION_DELAY);
    },
  },
};
