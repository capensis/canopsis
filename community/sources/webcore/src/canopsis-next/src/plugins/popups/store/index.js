import { POPUP_TYPES } from '@/constants';
import { POPUP_AUTO_CLOSE_DELAY } from '@/config';

import { uid } from '@/helpers/uid';

export const types = {
  ADD: 'ADD',
  REMOVE: 'REMOVE',
  SET_DEFAULT_CLOSE_TIME: 'SET_DEFAULT_CLOSE_TIME',
};

export default {
  namespaced: true,
  state: {
    defaultCloseTimeByType: {
      [POPUP_TYPES.error]: POPUP_AUTO_CLOSE_DELAY,
      [POPUP_TYPES.success]: POPUP_AUTO_CLOSE_DELAY,
      [POPUP_TYPES.info]: POPUP_AUTO_CLOSE_DELAY,
      [POPUP_TYPES.warning]: POPUP_AUTO_CLOSE_DELAY,
    },
    popups: [],
  },
  getters: {
    popups: state => state.popups,
  },
  mutations: {
    [types.ADD](state, { popup }) {
      state.popups.push(popup);
    },
    [types.REMOVE](state, { id }) {
      state.popups = state.popups.filter(v => v.id !== id);
    },
    [types.SET_DEFAULT_CLOSE_TIME](state, { type, time }) {
      state.defaultCloseTimeByType[type] = time;
    },
  },
  actions: {
    add({ commit, state }, {
      id = uid('popup'),
      type,
      text,
      autoClose = state.defaultCloseTimeByType[type],
    } = {}) {
      commit(types.ADD, {
        popup: {
          id, type, text, autoClose,
        },
      });
    },

    setDefaultCloseTime({ commit }, { type, time = POPUP_AUTO_CLOSE_DELAY }) {
      return commit(types.SET_DEFAULT_CLOSE_TIME, { type, time });
    },

    remove({ commit }, { id }) {
      commit(types.REMOVE, { id });
    },

    success({ dispatch }, popup) {
      return dispatch('add', { ...popup, type: POPUP_TYPES.success });
    },

    info({ dispatch }, popup) {
      return dispatch('add', { ...popup, type: POPUP_TYPES.info });
    },

    warning({ dispatch }, popup) {
      return dispatch('add', { ...popup, type: POPUP_TYPES.warning });
    },

    error({ dispatch }, popup) {
      return dispatch('add', { ...popup, type: POPUP_TYPES.error });
    },
  },
};
