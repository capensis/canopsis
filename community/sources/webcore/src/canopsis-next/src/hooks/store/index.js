import { useComponentInstance } from '../vue';

/**
 * Normalizes a given map into an array of objects with `key` and `value` properties.
 * If the input is an array, each element is treated as both the key and value.
 * If the input is an object, it converts the entries into an array of objects.
 *
 * @param {Object|Array} map - The map to normalize, which can be either an object or an array.
 * @returns {Array.<{key: string, value: string}>} An array of objects with `key` and `value` properties.
 */
const normalizeStoreMap = (map = {}) => (
  Array.isArray(map)
    ? map.map(key => ({ key, value: key }))
    : Object.entries(map).map(([key, value]) => ({ key, value }))
);

/**
 * Retrieves the Vuex store instance from the current Vue component instance.
 * This function must be used within the Vue component's `setup()` method.
 *
 * @returns {Object} The Vuex store instance.
 * @throws {Error} If the function is not used within the `setup()` method.
 *
 * @example
 * // In a Vue component
 * import { useStore } from `./path/to/useStore`;
 *
 * export default {
 *   setup() {
 *     const store = useStore();
 *     // Now you can use the store to commit mutations, dispatch actions, or access state
 *     return {
 *       // Your other setup properties
 *     };
 *   }
 * }
 */
export const useStore = () => {
  const vm = useComponentInstance();

  if (!vm) {
    throw new Error('You must use this function within the "setup()" method');
  }

  return vm.$store;
};

/**
 * Creates hooks for accessing Vuex store module's getters and actions.
 * This function helps in creating a more modular and organized way to interact with Vuex store modules.
 *
 * @param {string} namespace - The namespace of the Vuex store module.
 * @returns {Object} An object containing the store, module, and functions to access getters and actions.
 * @throws {Error} If the provided namespace does not correspond to any module in the store.
 *
 * @example
 * // Assuming a Vuex store module is registered under `user/` namespace
 * import { useStoreModuleHooks } from `./path/to/useStoreModuleHooks`;
 *
 * export default {
 *   setup() {
 *     const { useGetters, useActions } = useStoreModuleHooks('user/');
 *     const { userName } = useGetters(['userName']);
 *     const { updateUser } = useActions({ customUpdateUser: 'updateUser' });
 *
 *     return {
 *       userName,
 *       customUpdateUser
 *     };
 *   }
 * }
 */
export const useStoreModuleHooks = (namespace) => {
  const store = useStore();
  const preparedNamespace = namespace.slice(-1) === '/' ? namespace : `${namespace}/`;
  const module = store._modulesNamespaceMap[preparedNamespace];

  if (!module) {
    throw new Error(`Incorrect module namespace = ${namespace}`);
  }

  const useGetters = getters => normalizeStoreMap(getters).reduce((acc, { key, value }) => {
    acc[key] = () => store.getters(preparedNamespace + value);

    return acc;
  }, {});

  const useActions = actions => normalizeStoreMap(actions).reduce((acc, { key, value }) => {
    acc[key] = (...args) => module.context.dispatch.apply(store, [value, ...args]);

    return acc;
  }, {});

  return {
    store,
    module,
    useGetters,
    useActions,
  };
};
