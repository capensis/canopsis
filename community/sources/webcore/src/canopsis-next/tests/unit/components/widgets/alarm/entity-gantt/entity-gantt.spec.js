import { mount, shallowMount, createVueInstance } from '@unit/utils/vue';

import { createMockedStoreModules } from '@unit/utils/store';
import flushPromises from 'flush-promises';
import { mockPopups } from '@unit/utils/mock-hooks';
import EntityGantt from '@/components/widgets/alarm/entity-gantt/entity-gantt.vue';

const localVue = createVueInstance();

const stubs = {
  'c-progress-overlay': true,
  'junit-gantt-chart': true,
};

const factory = (options = {}) => shallowMount(EntityGantt, {
  localVue,
  stubs,

  ...options,
});

const snapshotFactory = (options = {}) => mount(EntityGantt, {
  localVue,
  stubs,

  ...options,
});

const selectJunitGanttChart = wrapper => wrapper.find('junit-gantt-chart-stub');

describe('entity-gantt', () => {
  const $popups = mockPopups();
  const fetchItemGanttIntervalsWithoutStore = jest.fn().mockResolvedValue({
    meta: {},
    data: [],
  });
  const testSuiteModule = {
    name: 'testSuite/entityGantt',
    actions: {
      fetchItemGanttIntervalsWithoutStore,
    },
  };
  const store = createMockedStoreModules([
    testSuiteModule,
  ]);

  const ganttIntervals = [{}, {}];
  const ganttIntervalsMeta = {
    total_count: 2,
  };
  const alarm = {
    entity: {
      _id: 'entity-id',
    },
  };

  afterEach(() => {
    fetchItemGanttIntervalsWithoutStore.mockClear();
  });

  it('Gantt intervals fetched after mount', async () => {
    const wrapper = factory({
      store,
      propsData: {
        alarm,
      },
    });

    const junitGanttChart = selectJunitGanttChart(wrapper);

    const newQuery = {
      rowsPerPage: 15,
      page: 2,
    };

    junitGanttChart.vm.$emit('update:query', newQuery);

    await flushPromises();

    expect(fetchItemGanttIntervalsWithoutStore).toBeCalledWith(
      expect.any(Object),
      {
        id: alarm.entity._id,
        params: {
          limit: newQuery.rowsPerPage,
          page: newQuery.page,
        },
      },
      undefined,
    );
  });

  it('Gantt intervals fetched after change the query', async () => {
    fetchItemGanttIntervalsWithoutStore.mockResolvedValueOnce({
      meta: ganttIntervalsMeta,
      data: ganttIntervals,
    });
    factory({
      store,
      propsData: {
        alarm,
      },
    });

    await flushPromises();

    expect(fetchItemGanttIntervalsWithoutStore).toBeCalledWith(
      expect.any(Object),
      {
        id: alarm.entity._id,
        params: {
          limit: 10,
          page: 1,
        },
      },
      undefined,
    );
  });

  it('Error popup showed after fetch gantt intervals with error message', async () => {
    const error = {
      message: 'Error message',
    };
    fetchItemGanttIntervalsWithoutStore.mockRejectedValueOnce(error);
    factory({
      store,
      propsData: {
        alarm,
      },
      mocks: {
        $popups,
      },
    });

    await flushPromises();

    expect($popups.error).toBeCalledWith({ text: error.message });
  });

  it('Error popup showed after fetch gantt intervals with error description', async () => {
    const error = {
      description: 'Error description',
    };
    fetchItemGanttIntervalsWithoutStore.mockRejectedValueOnce(error);
    factory({
      store,
      propsData: {
        alarm,
      },
      mocks: {
        $popups,
      },
    });

    await flushPromises();

    expect($popups.error).toBeCalledWith({ text: error.description });
  });

  it('Error popup showed after fetch gantt intervals with error', async () => {
    fetchItemGanttIntervalsWithoutStore.mockRejectedValueOnce({});
    factory({
      store,
      propsData: {
        alarm,
      },
      mocks: {
        $popups,
      },
    });

    await flushPromises();

    expect($popups.error).toBeCalledWith({ text: 'Something went wrong...' });
  });

  it('Renders `entity-gantt` with required props', async () => {
    fetchItemGanttIntervalsWithoutStore
      .mockResolvedValueOnce({
        meta: ganttIntervalsMeta,
        data: ganttIntervals,
      });
    const wrapper = snapshotFactory({
      store,
      propsData: {
        alarm,
      },
    });

    expect(wrapper.element).toMatchSnapshot();

    await flushPromises();

    expect(wrapper.element).toMatchSnapshot();
  });
});
