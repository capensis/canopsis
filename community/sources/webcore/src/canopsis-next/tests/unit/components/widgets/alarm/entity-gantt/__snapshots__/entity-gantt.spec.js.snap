// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity-gantt Renders \`entity-gantt\` with required props 1`] = `
<div
  class="layout justify-center"
>
  <c-progress-overlay-stub
    pending="true"
  />
  <junit-gantt-chart-stub
    items=""
    query="[object Object]"
    totalitems="0"
    width="840"
  />
</div>
`;

exports[`entity-gantt Renders \`entity-gantt\` with required props 2`] = `
<div
  class="layout justify-center"
>
  <c-progress-overlay-stub />
  <junit-gantt-chart-stub
    items="[object Object],[object Object]"
    query="[object Object]"
    totalitems="2"
    width="840"
  />
</div>
`;
